Sync stats migration

alovew opened the pull request:
Sync stats data is nested in a JSON field on the attempts table. This PR creates a new normalized table to store sync stats data so it is easily queryable.

jdpgrailsdev commented:
@alovew Can we just removed the commented out code entirely and just revert this PR if we need to get it back?

gosusnp reviewed (changes requested):
I am a bit confused here.
Are you modifying an old migration? Also from the look of schema_dump.txt, it feels like you did not add one migration to your change.

alovew commented:
@jdpgrailsdev and @gosusnp sorry I committed the wrong thing! I had to comment out a part of an old migration because it was causing migrations to fail, so I commented it in order to test running this new migration locally. Removed that from this PR and added in the actual migration.

jdpgrailsdev reviewed (approved):
:shipit:

gosusnp commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,68 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO: update migration description in the class name
```
Can remove the TODO.

gosusnp commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,68 @@
+        .constraints(unique(attemptId))
+        .execute();
+
+    ctx.createIndex("attempt_id_idx").on("sync_stats", "attempt_id").execute();
```
I suspect our attempts don't have a very wide range of value. What's the expected use case here?

alovew replied:
for the normalization ticket I was working on, we need to query by attempt_id in order to get stats for each previous attempt for a given job.

gosusnp commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  private static void createSyncStatsTable(final DSLContext ctx) {
+    final Field<UUID> id = DSL.field("id", SQLDataType.UUID.nullable(false));
```
Shouldn't we have a `jobId` as well?

alovew replied:
no, that is stored on the attempts table. the sync stats are associated with one attempt of a job.

gosusnp commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,68 @@
+        .columns(id, attemptId, recordsEmitted, bytesEmitted, sourceStateMessagesEmitted, destinationStateMessagesEmitted, recordsCommitted,
+            meanSecondsBeforeSourceStateMessageEmitted, maxSecondsBeforeSourceStateMessageEmitted, meanSecondsBetweenStateMessageEmittedandCommitted,
+            maxSecondsBetweenStateMessageEmittedandCommitted, createdAt, updatedAt)
+        .constraints(unique(attemptId))
```
Shouldn't `attemptId` be composed with some other id? I am thinking jobId for example.

gosusnp reviewed (changes requested):
My comments are probably tied to this question: is attemptId here the same as the attempts on temporal?
If they are indeed the same, I think they will collide across syncs so there should be another key.

gosusnp reviewed (approved):
Got it, `attemptId != attempt_number`.

gosusnp commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,68 @@
+        .columns(id, attemptId, recordsEmitted, bytesEmitted, sourceStateMessagesEmitted, destinationStateMessagesEmitted, recordsCommitted,
+            meanSecondsBeforeSourceStateMessageEmitted, maxSecondsBeforeSourceStateMessageEmitted, meanSecondsBetweenStateMessageEmittedandCommitted,
+            maxSecondsBetweenStateMessageEmittedandCommitted, createdAt, updatedAt)
+        .constraints(unique(attemptId))
```
Can we make this a foreign key instead? I think that would make things clearer.

alovew replied:
good call, updated 👍

alovew commented:
hmmm not sure why I'm seeing this error: `Caused by: org.postgresql.util.PSQLException: ERROR: relation "attempts" does not exist`

It's really similar to the error I was seeing locally from an earlier migration - that one was saying that airbyte_configs didn't exist, even though it seemed to me like it did.

evantahler commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_40_3_003__CreateSyncStats.java:
```diff
@@ -0,0 +1,69 @@
+        DSL.field("max_seconds_between_state_message_emitted_and_committed", SQLDataType.INTEGER.nullable(true));
+    final Field<OffsetDateTime> createdAt =
+        DSL.field("created_at", SQLDataType.TIMESTAMPWITHTIMEZONE.nullable(false).defaultValue(currentOffsetDateTime()));
+    final Field<OffsetDateTime> updatedAt =
+        DSL.field("updated_at", SQLDataType.TIMESTAMPWITHTIMEZONE.nullable(false).defaultValue(currentOffsetDateTime()));
```
Because I don't know - do we expect the default times to be set by the application when a new row is written, or do we want the database to have a default value of `NOW()`?

alovew replied:
hmm so I was using a default value the way some other migrations are under the assumption that these fields would be generated automatically without the application specifying the current time, but in our actual implementations I'm seeing that we are having the application set the current time for each row. So I'm not sure exactly how the default value works in this case.

alovew commented:
@gosusnp and @jdpgrailsdev I updated this PR to avoid the migration error I was seeing. for some reason when the sync stats migration is run, it is not detecting the `attempts` table so I can't write the foreign key correctly. Instead I'm just putting a unique constraint on it.

alovew commented:
@gosusnp and @jdpgrailsdev sorry for all the notifications! I had to make some changes to make this work. 

1. The attempts table is in the jobs db, so I moved the migration to there
2. when there is a foreign key association, truncate commands fail unless the truncation is cascaded to these associations, so I updated this

alovew commented on airbyte-scheduler/scheduler-persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:
```diff
@@ -802,7 +802,7 @@ private static void truncateTable(final DSLContext ctx, final String schema, fin
     ctx.dropTableIfExists(backupTableSql).execute();
     ctx.createTable(backupTableSql).as(DSL.select(DSL.asterisk()).from(tableSql)).withData().execute();
-    ctx.truncateTable(tableSql).restartIdentity().execute();
+    ctx.truncateTable(tableSql).restartIdentity().cascade().execute();
```
i'm a little concerned about this change, since this function seems to be truncating and backing up tables, but the associated rows that are deleted from the sync_stats table will not be backed up. It looks like this is called when exporting a db to a new location (possibly?) and this will work as long as the tables are moved in the correct order. Not sure if this is ok to live with for now or not.

alovew replied:
cc: @ChristopheDuong since it looks like you wrote most of this

ChristopheDuong replied:
The "backup" logic was introduced at a time when almost everything was persisted in a single table (json blobs?). But now that we are splitting more things out, cascading may make sense but the "backup" logic is probably deprecated...

IIRC backing things up was mainly due to being able to roll back if migrations were not successful, but again since this was written, the migration module/code also evolved.
